### PR TITLE
Fix big endian issues

### PR DESCRIPTION
**Pull request by ibc**
Working on fixing Big-Endian related issues described in issue #638.

Still WIP (although it may be done).

### CODE REVIEW

**Comment by nazar-pc** [comment]
Tests that were failing before do pass now, but media flow doesn't seem to work yet (not errors just like before)

**Comment by ibc** [comment]
> Tests that were failing before do pass now, but media flow doesn't seem to work yet (not errors just like before)

I don't think we are done yet. We must check whether we also need to consider Big-Endian in codec parsing.

**Comment by ibc** [comment]
@jmillan definitely I need help on this. I've pushed this. I've added a `headerSize = 8` to remove `reserved1` and `reserved2` and replace with a single `uint16_t reserved1`, and it still produces a SIGSEGV:

```diff
diff --git a/worker/include/RTC/RTCP/FeedbackPsTst.hpp b/worker/include/RTC/RTCP/FeedbackPsTst.hpp
index 0d2b424c8..1c307fb8e 100644
--- a/worker/include/RTC/RTCP/FeedbackPsTst.hpp
+++ b/worker/include/RTC/RTCP/FeedbackPsTst.hpp
@@ -28,19 +28,19 @@ namespace RTC
 			{
 				uint32_t ssrc;
 				uint8_t sequenceNumber;
-				uint8_t reserved1;
-				uint8_t reserved2;
+				uint16_t reserved1;
 #if defined(MS_LITTLE_ENDIAN)
 				uint8_t index : 5;
-				uint8_t reserved3 : 3;
+				uint8_t reserved2 : 3;
 #elif defined(MS_BIG_ENDIAN)
-				uint8_t reserved3 : 3;
+				uint8_t reserved2 : 3;
 				uint8_t index : 5;
 #endif
 			};
 
 		public:
 			static const FeedbackPs::MessageType messageType;
+			static const size_t headerSize;
 
 		public:
 			explicit FeedbackPsTstItem(Header* header) : header(header)
diff --git a/worker/src/RTC/RTCP/FeedbackPsTst.cpp b/worker/src/RTC/RTCP/FeedbackPsTst.cpp
index 6c08fc0a2..c3f931f54 100644
--- a/worker/src/RTC/RTCP/FeedbackPsTst.cpp
+++ b/worker/src/RTC/RTCP/FeedbackPsTst.cpp
@@ -14,11 +14,11 @@ namespace RTC
 		{
 			MS_TRACE();
 
-			this->raw    = new uint8_t[sizeof(Header)];
+			this->raw    = new uint8_t[FeedbackPsTstItem::headerSize];
 			this->header = reinterpret_cast<Header*>(this->raw);
 
 			// Set reserved bits to zero.
-			std::memset(this->header, 0, sizeof(Header));
+			std::memset(this->header, 0, FeedbackPsTstItem::headerSize);
 
 			this->header->ssrc           = uint32_t{ htonl(ssrc) };
 			this->header->sequenceNumber = sequenceNumber;
@@ -30,9 +30,9 @@ namespace RTC
 		{
 			MS_TRACE();
 
-			std::memcpy(buffer, this->header, sizeof(Header));
+			std::memcpy(buffer, this->header, FeedbackPsTstItem::headerSize);
 
-			return sizeof(Header);
+			return FeedbackPsTstItem::headerSize;
 		}
 
 		template<typename T>
@@ -52,12 +52,16 @@ namespace RTC
 		template<>
 		const FeedbackPs::MessageType FeedbackPsTstItem<Tstr>::messageType =
 		  FeedbackPs::MessageType::TSTR;
+		template<>
+		const size_t FeedbackPsTstItem<Tstr>::headerSize = 8;
 
 		/* Specialization for Tstn class. */
 
 		template<>
 		const FeedbackPs::MessageType FeedbackPsTstItem<Tstn>::messageType =
 		  FeedbackPs::MessageType::TSTN;
+		template<>
+		const size_t FeedbackPsTstItem<Tstn>::headerSize = 8;
 
 		// Explicit instantiation to have all definitions in this file.
 		template class FeedbackPsTstItem<Tstr>;
```

**Comment by ibc** [comment]
Good. So no more sizeof() in the whole code?

**Comment by jmillan** [comment]
> Good. So no more sizeof() in the whole code?

I mostly checked RTCP. Will check it further.

**Comment by jmillan** [comment]
We are done now with all `sizeof` to structs.

**Comment by ibc** [comment]
> We are done now with all `sizeof` to structs.

What about this?

```
RTC/RTCP/XrDelaySinceLastRr.cpp
19:			if (len < sizeof(Body))
52:			std::memcpy(buffer, this->body, sizeof(DelaySinceLastRr::SsrcInfo::Body));
54:			return sizeof(DelaySinceLastRr::SsrcInfo::Body);
69:			while (len > offset && reportLength >= sizeof(DelaySinceLastRr::SsrcInfo::Body))
85:				reportLength -= sizeof(DelaySinceLastRr::SsrcInfo::Body);
97:			size_t length = static_cast<uint16_t>((sizeof(SsrcInfo::Body) * this->ssrcInfos.size() / 4));
126:			  static_cast<uint16_t>((sizeof(SsrcInfo::Body) * this->ssrcInfos.size() / 4)));

RTC/RTCP/FeedbackPsRemb.cpp
126:			offset += sizeof(FeedbackPsRembPacket::uniqueIdentifier);

RTC/RTCP/XrReceiverReferenceTime.cpp
19:			if (len < ExtendedReportBlock::CommonHeaderSize + sizeof(ReceiverReferenceTime::Body))
60:			std::memcpy(buffer + offset, this->body, sizeof(ReceiverReferenceTime::Body));
62:			offset += sizeof(ReceiverReferenceTime::Body);
```

Actually such a `Body` is a struct:

```c++
struct Body
{
	uint32_t ssrc;
	uint32_t lrr;
	uint32_t dlrr;
};
```

**Comment by jmillan** [comment]
> What about this?

Will handle those

**Comment by ibc** [comment]
@nazar-pc in theory we are done here. Wondering if you can test this in that IBM Big-Endian machine 😀

**Comment by nazar-pc** [comment]
Tests are fine, media flow still doesn't work. I have enabled worker logs and saw a lot of these:
```
[2021-08-26T13:46:44Z DEBUG mediasoup::worker] [id:08e804f7-adac-4168-87a3-013f554dd3e2] RTC::SrtpSession::DecryptSrtp() | srtp_unprotect() failed: authentication failure (srtp_err_status_auth_fail)
[2021-08-26T13:46:44Z WARN  mediasoup::worker] [id:08e804f7-adac-4168-87a3-013f554dd3e2] RTC::WebRtcTransport::OnRtpDataReceived() | DecryptSrtp() failed [ssrc:4002646541, payloadType:120, seq:13479]
[2021-08-26T13:46:44Z DEBUG mediasoup::worker] [id:08e804f7-adac-4168-87a3-013f554dd3e2] RTC::SrtpSession::DecryptSrtp() | srtp_unprotect() failed: authentication failure (srtp_err_status_auth_fail)
[2021-08-26T13:46:44Z WARN  mediasoup::worker] [id:08e804f7-adac-4168-87a3-013f554dd3e2] RTC::WebRtcTransport::OnRtpDataReceived() | DecryptSrtp() failed [ssrc:2612743051, payloadType:109, seq:22167]
```

Also these occasionally:
```
[2021-08-26T13:46:45Z DEBUG mediasoup::worker] [id:08e804f7-adac-4168-87a3-013f554dd3e2] RTC::Producer::ReceiveRtcpSenderReport() | RtpStream not found [ssrc:4002646541]
```

**Comment by ibc** [comment]
@nazar-pc we are gonna merge this and create a separate issue for the `RTC::SrtpSession::DecryptSrtp() | srtp_unprotect() failed` error in Big-Endian.

File: worker/include/RTC/RTCP/FeedbackRtpNack.hpp [hunk 1]
```diff
@@ -28,6 +28,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t HeaderSize = 4;
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackRtpNack.hpp]:
```suggestion
			static const size_t HeaderSize{ 4 };
```

File: worker/include/RTC/RTCP/Feedback.hpp [hunk 2]
```diff
@@ -21,6 +21,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t HeaderSize = 8;
```
Comment (ibc) [inline worker/include/RTC/RTCP/Feedback.hpp]:
```suggestion
			static const size_t HeaderSize{ 8 };
```

File: worker/include/RTC/RTCP/FeedbackPsLei.hpp [hunk 3]
```diff
@@ -26,6 +26,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t HeaderSize = 4;
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackPsLei.hpp]:
```suggestion
			static const size_t HeaderSize{ 4 };
```

File: worker/include/RTC/RTCP/FeedbackPsSli.hpp [hunk 4]
```diff
@@ -26,6 +26,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t HeaderSize = 4;
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackPsSli.hpp]:
```suggestion
			static const size_t HeaderSize{ 4 };
```

File: worker/include/RTC/RTCP/FeedbackPsTst.hpp [hunk 5]
```diff
@@ -28,12 +27,19 @@ namespace RTC
...
+				uint8_t index : 5;
+				uint8_t reserved2 : 3;
+#elif defined(MS_BIG_ENDIAN)
+				uint8_t reserved2 : 3;
+				uint8_t index : 5;
+#endif
 			};
 
 		public:
+			static const size_t HeaderSize = 8;
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackPsTst.hpp]:
```suggestion
			static const size_t HeaderSize{ 8 };
```

File: worker/include/RTC/RTCP/FeedbackPsVbcm.hpp [hunk 6]
```diff
@@ -30,13 +30,19 @@ namespace RTC
...
+#elif defined(MS_BIG_ENDIAN)
 				uint8_t zero : 1;
 				uint8_t payloadType : 7;
+#endif
 				uint16_t length;
 				uint8_t value[];
 			};
 
 		public:
+			static const size_t HeaderSize = 8;
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackPsVbcm.hpp]:
```suggestion
			static const size_t HeaderSize{ 8 };
```

File: worker/include/RTC/RTCP/FeedbackRtpEcn.hpp [hunk 7]
```diff
@@ -42,6 +42,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t HeaderSize = 20;
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackRtpEcn.hpp]:
```suggestion
			static const size_t HeaderSize{ 20 };
```

File: worker/include/RTC/RTCP/FeedbackRtpTllei.hpp [hunk 8]
```diff
@@ -27,6 +27,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t HeaderSize = 4;
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackRtpTllei.hpp]:
```suggestion
			static const size_t HeaderSize{ 4 };
```

File: worker/include/RTC/RTCP/FeedbackRtpTmmb.hpp [hunk 9]
```diff
@@ -31,6 +31,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t HeaderSize = 8;
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackRtpTmmb.hpp]:
```suggestion
			static const size_t HeaderSize{ 8 };
```

File: worker/include/RTC/RTCP/Packet.hpp [hunk 10]
```diff
@@ -44,18 +44,19 @@ namespace RTC
...
 				uint8_t count : 5;
 #endif
-				uint8_t packetType : 8;
-				uint16_t length : 16;
+				uint8_t packetType;
+				uint16_t length;
 			};
 
 		public:
+			static const size_t CommonHeaderSize = 4;
```
Comment (ibc) [inline worker/include/RTC/RTCP/Packet.hpp]:
```suggestion
			static const size_t CommonHeaderSize{ 4 };
```

File: worker/include/RTC/RTCP/ReceiverReport.hpp [hunk 11]
```diff
@@ -26,6 +26,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t HeaderSize = 24;
```
Comment (ibc) [inline worker/include/RTC/RTCP/ReceiverReport.hpp]:
```suggestion
			static const size_t HeaderSize{ 24 };
```

File: worker/include/RTC/RTCP/SenderReport.hpp [hunk 12]
```diff
@@ -24,6 +24,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t HeaderSize = 24;
```
Comment (ibc) [inline worker/include/RTC/RTCP/SenderReport.hpp]:
```suggestion
			static const size_t HeaderSize{ 24 };
```

File: worker/include/RTC/RTCP/XR.hpp [hunk 13]
```diff
@@ -46,6 +46,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t CommonHeaderSize = 4;
```
Comment (ibc) [inline worker/include/RTC/RTCP/XR.hpp]:
```suggestion
			static const size_t CommonHeaderSize{ 4 };
```

File: worker/include/RTC/RTCP/XrDelaySinceLastRr.hpp [hunk 14]
```diff
@@ -36,6 +36,7 @@ namespace RTC
 			class SsrcInfo
 			{
 			public:
+				static const size_t BodySize = 12;
```
Comment (ibc) [inline worker/include/RTC/RTCP/XrDelaySinceLastRr.hpp]:
```suggestion
				static const size_t BodySize{ 12 };
```

File: worker/include/RTC/RTCP/XrReceiverReferenceTime.hpp [hunk 15]
```diff
@@ -31,6 +31,7 @@ namespace RTC
 			};
 
 		public:
+			static const size_t BodySize = 8;
```
Comment (ibc) [inline worker/include/RTC/RTCP/XrReceiverReferenceTime.hpp]:
```suggestion
			static const size_t BodySize{ 8 };
```

File: worker/include/RTC/RtpPacket.hpp [hunk 16]
```diff
@@ -114,6 +114,7 @@ namespace RTC
 		};
 
 	public:
+		static const size_t HeaderSize = 12;
```
Comment (ibc) [inline worker/include/RTC/RtpPacket.hpp]:
```suggestion
		static const size_t HeaderSize{ 12 };
```

**Comment by ibc** [comment]
Issue created: https://github.com/versatica/mediasoup/issues/656

**Comment by ibc** [comment]
Missing issue in libsrtp amy be fixed in Meson PR since it tells libsrtp about BigEnding:

https://github.com/cisco/libsrtp/blob/master/meson.build#L59-L61

**Pull request merged**